experimental/parser/parse_decl.go

```diff
@@ -0,0 +1,695 @@
+		case !path.Nil():
+			// This will catch someone writing `import foo.bar;` when we legalize.
+			args.ImportPath = ast.ExprPath{Path: path}.AsAny()
+		}
+
+		if args.ImportPath.Nil() && canStartExpr(next) {
+			args.ImportPath = parseExpr(p, c, in.In())
+		}
+
+		if next := c.Peek(); next.Text() == "[" && !next.IsLeaf() {
+			args.Options = parseOptions(p, c.Pop(), in)
+		}
+
+		semi, err := p.Punct(c, ";", in.After())
+		args.Semicolon = semi
+		if err != nil && args.ImportPath.Nil() {
+			p.Error(err)
+		}
+
+		return p.NewDeclImport(args).AsAny()
+
+	case "reserved", "extensions":
+		if next.Text() == "=" {
+			// If whatever follows the path is an =, we're going to assume this
+			// is trying to be a field.
+			break
+		}
+
+		// Otherwise, rewind the cursor to before we parsed a type, and
+		// parse a range instead. Rewinding is necessary because otherwise we get
+		// into an annoying situation where if we have e.g. reserved foo to bar;
+		// we have already consumed reserved foo, but we want to push foo
+		// through the expression machinery to get foo to bar as a single
+		// expression.
+		c.Rewind(mark)
+		return parseRange(p, c).AsAny()
+	}
+
+	return parseDef(p, c, kw, ty, path).AsAny()
+}
+
+// parseBody parses a ({}-delimited) body of declarations.
+func parseBody(p *parser, braces token.Token, in taxa.Subject) ast.DeclBody {
+	body := p.NewDeclBody(braces)
+
+	// Drain the contents of the body into it. Remember,
+	// parseDecl must always make progress if there is more to
+	// parse.
+	c := braces.Children()
+	for !c.Done() {
+		if next := parseDecl(p, c, in); !next.Nil() {
+			body.Append(next)
+		}
+	}
+
+	return body
+}
+
+// parseRange parses a reserved/extensions range.
+func parseRange(p *parser, c *token.Cursor) ast.DeclRange {
+	// Consume the keyword token.
+	kw := c.Pop()
+
+	in := taxa.Extensions
+	if kw.Text() == "reserved" {
+		in = taxa.Reserved
+	}
+
+	// Consume expressions until we hit a semicolon or the [ of a compact
+	// options. Note that this means that we do not parse reserved [1, 2, 3];
+	// "correctly".
+	var (
+		// badExpr keeps track of whether we exited the loop due to a parse
+		// error or because we hit ; or [ or EOF.
+		badExpr bool
+		exprs   []exprComma
+	)
+
+	// Parse expressions until we hit a semicolon or the [ of compact options.
+	elems := commas(p, c, true, func(c *token.Cursor) (ast.ExprAny, bool) {
+		next := c.Peek().Text()
+		if next == ";" || next == "[" {
+			return ast.ExprAny{}, false
+		}
+
+		expr := parseExpr(p, c, in.In())
+		badExpr = badExpr || expr.Nil()
+
+		return expr, !expr.Nil()
+	})
+	elems(func(expr ast.ExprAny, comma token.Token) bool {
+		exprs = append(exprs, exprComma{expr, comma})
+		return true
+	})
+
+	var options ast.CompactOptions
+	if next := c.Peek(); next.Text() == "[" && !next.IsLeaf() {
+		options = parseOptions(p, c.Pop(), in)
+	}
+
+	// Parse a semicolon, if possible.
+	semi, err := p.Punct(c, ";", in.After())
+	if err != nil && !badExpr {
```

Comment:
Yes, good point. And no, we don't need to consider errors from parseOptions, because we only enter that if we see a `[...]` token.

Reply:
> we don't need to consider errors from parseOptions, because we only enter that if we see a `[...]` token

That's not what I was saying. IIUC, the `badExpr` variable is there so that you don't require a semicolon following a mangled definition. But if you see a `[...]` token, then do you care if there was a bad expression _before_ that? So, like I initially stated, if you fall into the if condition block that calls `parseOptions` above, I think you would set `badExpr` to false. Does that make sense?